src/goto-instrument/contracts/contract_wrangler.cpp

```diff
@@ -0,0 +1,391 @@
+
+/// \file
+/// Parse and annotate contracts
+
+#include "contract_wrangler.h"
+
+#include <util/expr_iterator.h>
+#include <util/format_expr.h>
+#include <util/pointer_expr.h>
+#include <util/simplify_expr.h>
+#include <util/std_types.h>
+
+#include <goto-programs/remove_unused_functions.h>
+
+#include <ansi-c/ansi_c_parser.h>
+#include <ansi-c/ansi_c_typecheck.h>
+#include <goto-instrument/contracts/utils.h>
+
+contract_wranglert::contract_wranglert(
+  goto_modelt &goto_model,
+  const std::string &file_name,
+  messaget &log)
+  : ns(goto_model.symbol_table),
+    goto_model(goto_model),
+    symbol_table(goto_model.symbol_table),
+    goto_functions(goto_model.goto_functions),
+    log(log)
+{
+  jsont configuration;
+
+  if(parse_json(file_name, log.get_message_handler(), configuration))
+    throw deserialization_exceptiont(file_name + " is not a valid JSON file");
+
+  configure_functions(configuration);
+
+  // Mangle loop contract function by function.
+  // We match function with a regex. There should one and only one function
+  // with name matched by the regex.
+  for(const auto &function_entry : this->functions)
+  {
+    bool function_found = 0;
+
+    for(const auto &function : goto_model.goto_functions.function_map)
+    {
+      // We find a matched function.
+      if(std::regex_match(function.first.c_str(), function_entry.first))
+      {
+        if(function_found)
+          throw deserialization_exceptiont(
+            "function regex" + function_entry.second.regex_str +
+            " match more than one function");
+
+        function_found = 1;
+
+        // Mangle all loop contracts in the function.
+        for(const auto &loop_entry : function_entry.second.loop_contract)
+        {
+          mangle(loop_entry, function.first);
+        }
+      }
+    }
+
+    if(!function_found)
+      throw deserialization_exceptiont(
+        "function " + function_entry.second.regex_str + " match no function");
+  }
+}
+
+void contract_wranglert::add_builtin_pointer_function_symbol(
+  std::string function_name,
+  const unsigned num_of_args)
+{
+  // Already exist.
+  if(symbol_table.lookup(CPROVER_PREFIX + function_name) != nullptr)
+    return;
+
+  code_typet::parameterst parameters;
+  for(unsigned i = 0; i < num_of_args; i++)
+  {
+    parameters.emplace_back(pointer_typet(signedbv_typet(32), 32));
+  }
+  symbolt fun_symbol(
+    CPROVER_PREFIX + function_name, code_typet(parameters, empty_typet()), "c");
+  symbol_table.add(fun_symbol);
+}
+
+void contract_wranglert::substitute_symbol(
+  exprt &expr,
+  const std::map<std::string, exprt> &symbol_map)
+{
+  for(depth_iteratort it = expr.depth_begin(); it != expr.depth_end(); it++)
+  {
+    if(it->id() != ID_symbol)
+      continue;
+    for(auto &symbol_entry : symbol_map)
+    {
+      if(it->get(ID_identifier) == symbol_entry.first)
+      {
+        exprt symbol(symbol_entry.second);
+        it.mutate().swap(symbol);
+        break;
+      }
+    }
+  }
+}
```

Comment:
Can't you use `replace_symbolt`?

Reply:
`replace_symbolt` requires that the type of the symbol must match the type of the replacement, which is not the case here. The symbol parsed from the fake function has no type.

Reply:
Use `unchecked_replace_symbolt` and be happy? :-)